aptos-move/framework/aptos-framework/sources/code.move

```diff
@@ -0,0 +1,183 @@
+        upgrade_policy: UpgradePolicy,
+        /// The list of modules installed by this package.
+        modules: vector<ModuleMetadata>,
+        /// Address aliases which where used when the modules above were compiled.
+        address_aliases: vector<AddressAlias>,
+        /// Dependencies which were used when this package was compiled.
+        deps: vector<PackageDep>,
+    }
+
+    /// Metadata about a module in a package.
+    struct ModuleMetadata has store, copy, drop {
+        /// Name of the module.
+        name: String,
+        /// Source map, in internal encoding
+        source_map: vector<u8>,
+        /// Source text.
+        source: String,
+    }
+
+    /// A package dependency. This consists of an address and the name of a package
+    /// at this address.
+    struct PackageDep has store, copy, drop {
+        addr: address,
+        name: String,
+    }
+
+    /// An address alias.
+    struct AddressAlias has store, copy, drop {
+        alias: String,
+        addr: address
+    }
+
+    /// Describes an upgrade policy
+    struct UpgradePolicy has store, copy, drop {
+        policy: u8
+    }
+
+    /// A package is attempted to publish with module names clashing with modules published by other packages on this
+    /// address.
+    const EMODULE_NAME_CLASH: u64 = 0x1;
+
+    /// A package is attempted to upgrade which is marked as immutable.
+    const EUPGRADE_IMMUTABLE: u64 = 0x2;
+
+    /// A package is attempted to upgrade with a weaker policy than previously.
+    const EUPGRADE_WEAKER_POLICY: u64 = 0x3;
+
+    /// Whether unconditional code upgrade with no compatibility check is allowed. This
+    /// publication mode should only be used for modules which aren't shared with user others.
+    /// The developer is responsible for not breaking memory layout of any resources he already
+    /// stored on chain.
+    public fun upgrade_policy_no_compat(): UpgradePolicy {
+        UpgradePolicy{policy: 0}
+    }
+
+    /// Whether a compatibility check should be performed for upgrades. The check only passes if
+    /// a new module has (a) the same public functions (b) for existing resources, no layout change.
+    public fun upgrade_policy_compat(): UpgradePolicy {
+        UpgradePolicy{policy: 1}
+    }
+
+    /// Whether the modules in the package are immutable and cannot be upgraded.
+    public fun upgrade_policy_immutable(): UpgradePolicy {
+        UpgradePolicy{policy: 2}
+    }
+
+    /// Whether the upgrade policy can be changed. In general, the policy can be only
+    /// strengthened but not weakened.
+    public fun can_change_upgrade_policy_to(from: UpgradePolicy, to: UpgradePolicy): bool {
+        from.policy <= to.policy
+    }
+
+    /// Publishes a package at the given signer's address. The caller must provide package metadata describing the
+    /// package.
+    public fun publish_package(owner: &signer, pack: PackageMetadata, code: vector<vector<u8>>) acquires PackageRegistry {
```

Comment:
Either I missed it or we'll need this module to expose a function to create PackageMetadata for scripts or modules to be able to call this function. The other function, publish_package_tx can pass bytes but this one is currently not callable because PackageMetadata cannot be created outside of this module

Reply:
Yes, working on this... Related to this the slack question about Rust <-> Move types.

Reply:
We still need to expose this function

Reply:
Not sure what you mean? There is now a 2nd function `public entry publish_package_txn`. How its used can be seen in the tests. There is also a Rust equivalent of `PackageMetadata` which allows to create this data on Rust client side.

Reply:
I'm aware of the separate publish_package_txn fn. But this publish_package is a public function (not public entry) and it takes PackageMetadata. This means this function can be called by another module or inside a script, but PackageMetadata cannot be created outside of this module. So how can other modules or scripts that call publish_package create and pass PackageMetadata?